src/db/package_version.rs

```diff
@@ -0,0 +1,168 @@
+use super::error::DatabaseError;
+use super::{models, schema, DbConn};
+use crate::api::publish::PublishRequest;
+use crate::models::{ApiToken, PackagePreview};
+use diesel::prelude::*;
+use uuid::Uuid;
+
+impl DbConn {
+    /// Insert a package version into the database and return the package version.
+    /// If the package doesn't exist, insert the package as well.
+    pub fn new_package_version(
+        &mut self,
+        api_token: &ApiToken,
+        request: &PublishRequest,
+    ) -> Result<models::PackageVersion, DatabaseError> {
+        // Check if the package exists.
+        let pkg_name = request.package_name.clone();
+        let package = if let Some(existing_package) = schema::packages::table
+            .filter(schema::packages::package_name.eq(pkg_name.clone()))
+            .select(schema::packages::all_columns)
+            .first::<models::Package>(self.inner())
+            .optional()
+            .map_err(|err| DatabaseError::QueryFailed(pkg_name.clone(), err))?
+        {
+            if existing_package.user_owner != api_token.user_id {
+                // The package exists but the user is not the owner.
+                return Err(DatabaseError::InvalidPublishToken);
+            }
+            Ok(existing_package)
+        } else {
+            // Insert a new package.
+            let new_package: models::NewPackage = models::NewPackage {
+                user_owner: api_token.user_id,
+                package_name: pkg_name.clone(),
+            };
+
+            let saved_package = diesel::insert_into(schema::packages::table)
+                .values(&new_package)
+                .returning(models::Package::as_returning())
+                .get_result(self.inner())
+                .map_err(|err| DatabaseError::InsertPackageFailed(pkg_name.clone(), err))?;
+            Ok(saved_package)
+        }?;
+
+        // Insert a new package version.
+        let new_version = models::NewPackageVersion {
```

Comment:
do we need some sort of date/time param here; would be good to have no?

Reply:
Yep, `created_at` has a default value that is set when the row is created. You can see it in the SQL migration file